File: guides/space-fighter-metal/src/Sources/SpaceFighter/Mesh.swift
import simd

/// How a mesh's vertices are assembled by the GPU.
enum Primitive {
    case triangle   // solid surfaces (ship, enemies, bullets)
    case line       // the ground grid
    case point      // the starfield
}

/// A CPU-side mesh: raw geometry the `Renderer` uploads into GPU buffers once at
/// startup. If `indices` is empty the mesh is drawn non-indexed.
struct Mesh {
    var vertices: [Vertex]
    var indices: [UInt16]
    var primitive: Primitive
}

// MARK: - Procedural geometry
//
// Everything here is generated in code — no model files, no asset pipeline.
// That's deliberate: "simple geometry for now". Each solid shape is *flat
// shaded*: every triangle gets its own three vertices and a single face normal,
// which gives the crisp faceted look of classic low-poly space shooters and
// keeps the maths trivial. Chapter 06 walks through this.

enum MeshLibrary {

    // MARK: Triangle-soup helpers

    /// Build a flat-shaded mesh from a list of triangles. Each triangle becomes
    /// three fresh vertices sharing one normal. The normal is computed from the
    /// winding, then flipped if it points *toward* the shape's centre, so faces
    /// always light correctly regardless of the order we listed corners in.
    private static func flat(_ tris: [(Vec3, Vec3, Vec3)]) -> Mesh {
        var sum = Vec3.zero
        for t in tris { sum += t.0 + t.1 + t.2 }
        let centroid = sum / Float(tris.count * 3)

        var vertices: [Vertex] = []
        var indices: [UInt16] = []
        vertices.reserveCapacity(tris.count * 3)
        indices.reserveCapacity(tris.count * 3)

        for t in tris {
            let faceCenter = (t.0 + t.1 + t.2) / 3
            var normal = simd_normalize(simd_cross(t.1 - t.0, t.2 - t.0))
            if simd_dot(normal, faceCenter - centroid) < 0 { normal = -normal }
            let base = UInt16(vertices.count)
            vertices.append(Vertex(position: t.0, normal: normal))
            vertices.append(Vertex(position: t.1, normal: normal))
            vertices.append(Vertex(position: t.2, normal: normal))
            indices.append(base)
            indices.append(base + 1)
            indices.append(base + 2)
        }
        return Mesh(vertices: vertices, indices: indices, primitive: .triangle)
    }

    /// Split a quad (given CCW) into two triangles.
    private static func quad(_ a: Vec3, _ b: Vec3, _ c: Vec3, _ d: Vec3) -> [(Vec3, Vec3, Vec3)] {
        [(a, b, c), (a, c, d)]
    }

    // MARK: Meshes

    /// The player's fighter: a sleek four-faced dart whose nose points down -Z
    /// (the forward convention used everywhere in this project).
    static func ship() -> Mesh {
        let nose  = Vec3( 0.0,  0.0, -2.0)
        let top   = Vec3( 0.0,  0.4,  1.0)
        let left  = Vec3(-1.3, -0.3,  1.0)
        let right = Vec3( 1.3, -0.3,  1.0)
        return flat([
            (nose, left, top),    // port hull
            (nose, top, right),   // starboard hull
            (nose, right, left),  // belly
            (top, left, right),   // engine deck (back)
        ])
    }

    /// A tumbling octahedron enemy — the simplest solid that still reads as a
    /// deliberate, hostile shape.
    static func enemy() -> Mesh {
        let r: Float = 0.9
        let px = Vec3(r, 0, 0), nx = Vec3(-r, 0, 0)
        let py = Vec3(0, r, 0), ny = Vec3(0, -r, 0)
        let pz = Vec3(0, 0, r), nz = Vec3(0, 0, -r)
        return flat([
            (py, px, pz), (py, pz, nx), (py, nx, nz), (py, nz, px),  // top cap
            (ny, pz, px), (ny, nx, pz), (ny, nz, nx), (ny, px, nz),  // bottom cap
        ])
    }

    /// A small cube used for the player's bolts (rendered unlit so it glows).
    static func projectile() -> Mesh {
        let h: Float = 0.5
        let v = [
            Vec3(-h, -h, -h), Vec3(h, -h, -h), Vec3(h, h, -h), Vec3(-h, h, -h),  // 0..3 back
            Vec3(-h, -h,  h), Vec3(h, -h,  h), Vec3(h, h,  h), Vec3(-h, h,  h),  // 4..7 front
        ]
        var tris: [(Vec3, Vec3, Vec3)] = []
        tris += quad(v[4], v[5], v[6], v[7])  // +Z
        tris += quad(v[1], v[0], v[3], v[2])  // -Z
        tris += quad(v[0], v[4], v[7], v[3])  // -X
        tris += quad(v[5], v[1], v[2], v[6])  // +X
        tris += quad(v[3], v[7], v[6], v[2])  // +Y
        tris += quad(v[0], v[1], v[5], v[4])  // -Y
        return flat(tris)
    }

    /// A cloud of point stars in a cube. The star vertex shader wraps these
    /// around the camera so the same few thousand points tile the whole sky
    /// (chapter 06 explains the modulo trick).
    static func starfield(count: Int, span: Float) -> Mesh {
        var vertices: [Vertex] = []
        vertices.reserveCapacity(count)
        let half = span * 0.5
        for _ in 0..<count {
            let p = Vec3(
                Float.random(in: -half...half),
                Float.random(in: -half...half),
                Float.random(in: -half...half)
            )
            // Vary brightness a touch via the normal.x channel, reused as a
            // scalar in the star shader.
            vertices.append(Vertex(position: p, normal: Vec3(Float.random(in: 0.4...1.0), 0, 0)))
        }
        return Mesh(vertices: vertices, indices: [], primitive: .point)
    }

    /// A flat reference grid on the XZ plane, drawn as lines. The renderer
    /// snaps it under the player so it reads as an endless floor/horizon.
    static func grid(halfExtent: Float, spacing: Float) -> Mesh {
        var vertices: [Vertex] = []
        var i = -halfExtent
        while i <= halfExtent + 0.001 {
            // Lines parallel to Z.
            vertices.append(Vertex(position: Vec3(i, 0, -halfExtent), normal: .zero))
            vertices.append(Vertex(position: Vec3(i, 0,  halfExtent), normal: .zero))
            // Lines parallel to X.
            vertices.append(Vertex(position: Vec3(-halfExtent, 0, i), normal: .zero))
            vertices.append(Vertex(position: Vec3( halfExtent, 0, i), normal: .zero))
            i += spacing
        }
        return Mesh(vertices: vertices, indices: [], primitive: .line)
    }
}

File: work/inputs/canonical/src/Mesh.cpp
#include "Mesh.hpp"

#include <array>
#include <random>

namespace {

using Tri = std::array<glm::vec3, 3>;

/// Flat-shade a triangle soup: every triangle gets its own three vertices and
/// one shared face normal, flipped outward if it points back at the centroid.
MeshData flat(const std::vector<Tri>& tris) {
    glm::vec3 centroid{0.0f};
    for (const Tri& t : tris) centroid += t[0] + t[1] + t[2];
    centroid /= float(tris.size() * 3);

    MeshData m;
    for (const Tri& t : tris) {
        glm::vec3 faceCenter = (t[0] + t[1] + t[2]) / 3.0f;
        glm::vec3 normal = glm::normalize(glm::cross(t[1] - t[0], t[2] - t[0]));
        if (glm::dot(normal, faceCenter - centroid) < 0.0f) normal = -normal;

        uint16_t base = uint16_t(m.vertices.size());
        for (const glm::vec3& p : t) m.vertices.push_back({p, normal});
        m.indices.insert(m.indices.end(), {base, uint16_t(base + 1), uint16_t(base + 2)});
    }
    return m;
}

void quad(std::vector<Tri>& out, glm::vec3 a, glm::vec3 b, glm::vec3 c, glm::vec3 d) {
    out.push_back({a, b, c});
    out.push_back({a, c, d});
}

} // namespace

namespace MeshLibrary {

MeshData ship() {
    glm::vec3 nose{0.0f, 0.0f, -2.0f};
    glm::vec3 top{0.0f, 0.4f, 1.0f};
    glm::vec3 left{-1.3f, -0.3f, 1.0f};
    glm::vec3 right{1.3f, -0.3f, 1.0f};
    return flat({
        {nose, left, top},
        {nose, top, right},
        {nose, right, left},
        {top, left, right},
    });
}

MeshData enemy() {
    glm::vec3 px{1.0f, 0.0f, 0.0f}, nx{-1.0f, 0.0f, 0.0f};
    glm::vec3 py{0.0f, 1.0f, 0.0f}, ny{0.0f, -1.0f, 0.0f};
    glm::vec3 pz{0.0f, 0.0f, 1.0f}, nz{0.0f, 0.0f, -1.0f};
    return flat({
        {py, px, pz}, {py, pz, nx}, {py, nx, nz}, {py, nz, px},
        {ny, pz, px}, {ny, nx, pz}, {ny, nz, nx}, {ny, px, nz},
    });
}

MeshData bolt() {
    glm::vec3 a{-1, -1, -1}, b{1, -1, -1}, c{1, 1, -1}, d{-1, 1, -1};
    glm::vec3 e{-1, -1, 1}, f{1, -1, 1}, g{1, 1, 1}, h{-1, 1, 1};
    std::vector<Tri> tris;
    quad(tris, a, b, c, d);
    quad(tris, f, e, h, g);
    quad(tris, e, a, d, h);
    quad(tris, b, f, g, c);
    quad(tris, d, c, g, h);
    quad(tris, e, f, b, a);
    return flat(tris);
}

MeshData starfield(int count, float span) {
    MeshData m;
    std::mt19937 rng{1337u};
    std::uniform_real_distribution<float> pos{-span * 0.5f, span * 0.5f};
    std::uniform_real_distribution<float> unit{0.0f, 1.0f};
    for (int i = 0; i < count; i++) {
        glm::vec3 p{pos(rng), pos(rng), pos(rng)};
        float brightness = 0.4f + 0.6f * unit(rng);
        m.vertices.push_back({p, glm::vec3(brightness, 0.0f, 0.0f)});
    }
    return m;
}

MeshData grid(float halfExtent, float spacing) {
    MeshData m;
    // One short segment per cell rather than one full-length line per row.
    // A single line running from far ahead of the camera to far behind it has
    // to be clipped against the near plane, and such lines can vanish
    // entirely; a one-cell segment never spans the camera.
    for (float i = -halfExtent; i <= halfExtent; i += spacing) {
        for (float j = -halfExtent; j < halfExtent; j += spacing) {
            m.vertices.push_back({{i, 0.0f, j}, {}});
            m.vertices.push_back({{i, 0.0f, j + spacing}, {}});
            m.vertices.push_back({{j, 0.0f, i}, {}});
            m.vertices.push_back({{j + spacing, 0.0f, i}, {}});
        }
    }
    return m;
}

} // namespace MeshLibrary
